### PR TITLE
SpamChecker metrics

### PR DESCRIPTION
**Pull request by jesopo**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by jesopo** [comment]
docstring is missing `Added in Synapse v1.2.3.` but I wasn't sure what's meant to go there for this one

**Comment by jesopo** [comment]
an obvious consideration is whether ModuleApi wants to superimpose the module's name in to the `Measure`'s `name`

**Comment by clokep** [comment]
From discussion on Matrix the goal of this is to be able to measure the amount of time a callback takes -- I wonder if it would make more sense to wrap some (or all) of the spots we call into modules with `Measure` blocks, tagging them with some sort of sensible name?

**Comment by jesopo** [comment]
I think that's a good idea so long as the performance impact of `Measure`s are negligible

**Comment by clokep** [comment]
> From discussion on Matrix the goal of this is to be able to measure the amount of time a callback takes -- I wonder if it would make more sense to wrap some (or all) of the spots we call into modules with `Measure` blocks, tagging them with some sort of sensible name?

There seemed to be general consensus in #synapse-dev that this makes sense to do.

**Comment by anoadragon453** [comment]
I'm taking the above as this PR requires further changes, and thus have added the Awaiting Changes label and am removing the request for review. But please re-request review it if that's incorrect!

**Comment by jesopo** [comment]
> From discussion on Matrix the goal of this is to be able to measure the amount of time a callback takes -- I wonder if it would make more sense to wrap some (or all) of the spots we call into modules with `Measure` blocks, tagging them with some sort of sensible name?

given the callbacks here are handled in `SpamChecker`, do I want to be putting the measurements in `SpamChecker` or in the things calling `SpamChecker`, or?

**Comment by DMRobertson** [comment]
> > From discussion on Matrix the goal of this is to be able to measure the amount of time a callback takes -- I wonder if it would make more sense to wrap some (or all) of the spots we call into modules with `Measure` blocks, tagging them with some sort of sensible name?
> 
> given the callbacks here are handled in `SpamChecker`, do I want to be putting the measurements in `SpamChecker` or in the things calling `SpamChecker`, or?

I'm not fully sure how `SpamChecker` relates to the module API, but given the choice I'd vote for putting them in the SpamChecker methods themselves.

**Comment by jesopo** [comment]
trying to think of a way to give each callback its own reasonable name to log the measurement unde...

```
>>> repr(A().a)
'<bound method A.a of <__main__.A object at 0x7fb1af964d00>>'
```
but this feels worse:
```
>>> A().a.__self__.__class__.__name__
'A'
>>> A().a.__name__
'a'
```
but I also don't think that `SpamChecker` has access to the names modules were loaded under, so that'll, I assume, want something like an additional `name` param for `SpamChecker.register_callbacks`

**Comment by squahtx** [comment]
> trying to think of a way to give each callback its own reasonable name to log the measurement under...
> 
> ```
> >>> repr(A().a)
> '<bound method A.a of <__main__.A object at 0x7fb1af964d00>>'
> ```
> 
> is bad but this feels worse:
> 
> ```
> >>> A().a.__self__.__class__.__name__
> 'A'
> >>> A().a.__name__
> 'a'
> ```
> 
> but I also don't think that `SpamChecker` has access to the names modules were loaded under, so that'll, I assume, want something like an additional `name` param for `SpamChecker.register_callbacks`

You could try `f"{method.__module__}.{method.__qualname__}"` ?

```py
>>> class SomeClass:
...      def instance_method(self):
...          pass
...
>>> SomeClass().instance_method.__qualname__
'SomeClass.instance_method'
```
```py
>>> from synapse.util import log_failure
>>> log_failure.__module__
'synapse.util'
>>> log_failure.__qualname__
'log_failure'
```

File: changelog.d/12513.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Insert Measure blocks throughout SpamChecker to collect metrics.
```
Comment (reivilibre) [inline changelog.d/12513.feature:1]: I'd just be tempted to make this a bit more comprehensible to an administrator (I don't know if it's clear what a Measure block is to anyone except Synapse devs)..

```suggestion
Measure the time taken in spam checker callbacks and expose them as per-block metrics in Prometheus.
```

I'm no expert but I think that assumes a bit less Synapse code knowledge?

**Review by reivilibre — APPROVED** [review]
> only bit I'm unsure of now is whether we need each Measure here to explicitly say what event it was handling, but that ought to be evident from the __module__.__qualname__s, so feels perhaps unnecessary

I agree — I don't think module authors are likely to reuse the same methods for different kinds of events so seems good to me!

File: changelog.d/12513.feature (hunk 1 above)
Comment (jesopo) [reply changelog.d/12513.feature]: i went a bit freestyle, but does the new version look OK?

Comment (reivilibre) [reply changelog.d/12513.feature]: Ya, works for me!

**Pull request merged**